src/samplers/resampler.jl

```diff
@@ -0,0 +1,53 @@
+struct WeightedResampler{F<:VariateForm, S<:ValueSupport} <: Sampleable{F, S}
+    obs::AbstractArray
+    wv::AbstractWeights
+end
+
+"""
+    WeightedResampler(obs::AbstractArray, wv::AbstractWeights)
+
+A WeightedResampler is a subtype of Distributions.Sampleable which randomly selects
+observations from the raw input data (`obs`) based on the weights (`wv`) provided.
+
+This type supports univariate, multivariate and matrixvariate forms, so `obs` can
+be a vector of values, matrix of values or a vector of matrices.
+"""
+function WeightedResampler(obs::T, wv::AbstractWeights) where T<:AbstractArray
+    F = _variate_form(T)
+    S = _value_support(eltype(T))
+
+    _validate(obs, wv)
+    WeightedResampler{F, S}(obs, wv)
+end
+
+_variate_form(::Type{<:AbstractVector}) = Univariate
```

Comment:
why not use `variate_form` and `value_support` instead of underscore functions?

Reply:
... because i didn't realize those functions already existed? :+1

Reply:
Would it make sense to document those functions in this MR? https://github.com/JuliaStats/Distributions.jl/pull/893/files

Reply:
yes indeed, let me fix that

Reply:
@rofinn I am thinking it is not a good idea to use these functions here, variate_form and value_support are supposed to take distributions in entry, passing them Array types would be some very unexpected use for them, so keeping the underscore functions is an acceptable solution here I think

Reply:
So we shouldn't extend those functions to support passing in an array type? Seems like something other samplers and distributions might find useful?

Reply:
It can lead to ambiguous behavior, we can keep this outside the scope of this PR.
Can you add some tests for the underscore functions? I know it is trivial, but we'll try to maintain the codecov high

Reply:
Sure. I believe all of those methods should already be getting hit, but more unit tests are always good 👍